### PR TITLE
atlas/schema: deploy changes with schema plan

### PR DESCRIPTION
**Pull request by giautm**

### CODE REVIEW

File: internal/controller/atlasschema_controller.go [hunk 1]
```diff
@@ -187,21 +187,131 @@ func (r *AtlasSchemaReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
+		// Try to apply the schema changes with lint policies
+		report, err = cli.SchemaApply(ctx, &atlasexec.SchemaApplyParams{
+			Env:         data.EnvName,
+			Vars:        vars,
+			To:          desiredURL,
+			TxMode:      string(data.TxMode),
+			AutoApprove: true,
+		})
+		// TODO: Better error handling for rejected changes.
+		if err != nil && strings.HasPrefix(err.Error(), "Rejected by review policy") {
```
Comment (giautm) [inline internal/controller/atlasschema_controller.go:213]: cc @a8m

File: internal/controller/atlasschema_controller.go [hunk 2]
```diff
@@ -187,21 +188,144 @@ func (r *AtlasSchemaReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
+		report, err = cli.SchemaApply(ctx, params)
+		// TODO: Better error handling for rejected changes.
+		if err != nil && strings.HasPrefix(err.Error(), "Rejected by review policy") {
+			log.Info("schema changes are rejected by the review policy, creating a new schema plan")
+			// The desired state is a file.
+			if data.Desired.Scheme == dbv1alpha1.SchemaTypeFile {
+				log.Info("schema is a file, pushing the schema to Atlas Cloud")
+				// We need to push the schema to the Atlas Cloud.
+				// This is to ensure that the schema is in sync with the Atlas Cloud.
+				temp, err := cli.SchemaPush(ctx, &atlasexec.SchemaPushParams{
```
Comment (rotemtam) [inline internal/controller/atlasschema_controller.go:272]: Need to verify this doesn't accidentally cause a situation where an operator loops and generates tons of junk on the server.

Comment (rotemtam) [reply internal/controller/atlasschema_controller.go:272]: genius

File: internal/controller/atlasschema_controller.go [hunk 3]
```diff
@@ -187,21 +188,149 @@ func (r *AtlasSchemaReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
+					Env:  data.EnvName,
+					Vars: vars,
+					Name: data.Cloud.Repo,
+					Tag:  fmt.Sprintf("operator-plan-%d", res.Generation),
```
[the remark this answers is not in the document]
Comment (rotemtam) [reply internal/controller/atlasschema_controller.go:278]: is the generation incrementally increasing or is it a ts?

[the remark this answers is not in the document]
Comment (rotemtam) [reply internal/controller/atlasschema_controller.go:278]: same resource with the same name can be created on separate clusters (dev/prod)

[the remark this answers is not in the document]
Comment (rotemtam) [reply internal/controller/atlasschema_controller.go:278]: maybe a hash of the plan itself?

**Review by rotemtam — APPROVED** [review]

**Pull request merged**